java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java

```diff
@@ -70,36 +71,149 @@ public void compositeColocateIdUsage(List<JsonObject> composites) {
     }
 
+    /**
+     * This method relocates {@code HashAdder} to the front of its parent
+     * $Unparellel$ which enables cat's-cradle shuffle. There are four
+     * scenarios.
+     * <p>
+     * Scenario 1 (optimized):
+     * <pre><code>
+     *     TStream<T> u = someStream.endParallel();
+     *     TStream<T> p = u.parallel(()->3, Routing.HASH_PARTITIONED);
+     * </code></pre>
+     * <BR>
+     * The code above demonstrates the simplest case where HashAdder is
+     * $Unparellel$'s only child, and $Unparallel$ is HashAdder's only
+     * parent:
+     * <pre><code>
+     *  $Unparallel$ -> HashAdder -> $Parallel$ -> HashRemover
+     * </code></pre>
+     * <BR>
+     * To enable cat's-cradle shuffle, this method moves {@code HashAdder}
+     * to the front of $Unparellel$.
+     * </p>
+     * <p>
+     * Scenario 2 (not yet optimized):
+     * <pre><code>
+     *     TStream<T> u1 = stream1.endParallel();
+     *     TStream<T> u2 = stream2.endParallel();
+     *     TStream<T> d = stream3.sample(0.5);
+     *     TStream<T> u = u1.union(u2);
+     *     TStream<T> p = u.parallel(()->3, Routing.HASH_PARTITIONED);
+     * </code></pre>
+     * <BR>
+     * This is slightly more complicated case where the {@code HashAdder}
+     * has multiple parents and each parent has only one child.
+     * <pre><code>
+     *           $Unparellel1$
+     *                |
+     *                V
+     *  sample -> HashAdder -> $Parallel$ -> HashRemover
+     *                ^
+     *                |
+     *           $Unparellel2$
+     *
+     * </code></pre>
+     * <BR>
+     * To enable cat's-cradle shuffle, this method inserts one copy of
+     * {@code HashAdder} to the front of every $Unparellel$, inserts one copy
+     * of {@code HashAdder} after every non-parallel parent, and removes the
+     * original {@code HashAdder}. The original graph will be modified to the
+     * following structure.
+     * <pre><code>
+     *           HashAdder -> $Unparellel1$
+     *                             |
+     *                             V
+     *  sample -> HashAdder -> $Parallel$ -> HashRemover
+     *                             ^
+     *                             |
+     *           HashAdder -> $Unparellel2$
+     * </code></pre>
+     * </p>
+     * <p>
+     * Scenario 3 (optimized):
+     * <pre><code>
+     *     TStream<T> u = input.endParallel();
+     *     TStream<T> p1 = u.parallel(()->3, keyer1);
+     *     TStream<T> p2 = u.parallel(()->4, keyer2);
+     *     TStream<T> f1 = u.filter((T x) -> true);
+     *     TStream<T> f2 = u.filter((T x) -> false);
+     * </code></pre>
+     * In this scenario, one $Unparallel$ can have multiple children, but each
+     * {@code HashAdder} has only one parent. The example code above results
+     * in the following graph structure.
+     * <pre><code>
+     *           HashAdder1 -> $Parellel1$ -> HashRemover1
```

Comment:
These diagrams would be clearer if one could relate `HashAdder1` to the code instantly, rather than having to guess/figure out which hash adder it is. Ie. it's the hash adder for `p1` with `keyer1`